sycl/include/sycl/ext/oneapi/work_group_static.hpp

```diff
@@ -0,0 +1,134 @@
+//==----- work_group_static.hpp --- SYCL group local memory extension -----==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#pragma once
+
+#include <sycl/access/access.hpp>             // for address_space, decorated
+#include <sycl/detail/defines_elementary.hpp> // for __SYCL_ALWAYS_INLINE
+#include <sycl/exception.hpp>                 // for exception
+#include <sycl/ext/intel/usm_pointers.hpp>    // for multi_ptr
+#include <sycl/ext/oneapi/properties/properties.hpp> // for properties
+
+#include <type_traits> // for enable_if_t
+
+namespace sycl {
+inline namespace _V1 {
+namespace ext::oneapi {
+namespace experimental {
+
+#ifdef __SYCL_DEVICE_ONLY__
+#define __SYCL_WG_SCOPE [[__sycl_detail__::wg_scope]]
+#else
+#define __SYCL_WG_SCOPE
+#endif
+
+/// @brief Allocate data in device local memory.
+/// Any work_group_static object will be place in device local memory and hold
+/// an object of type T. work_group_static object are implicitly treated as
+/// static.
+/// @tparam T must be a trivially constructible and destructible type
+template <typename T> class __SYCL_WG_SCOPE work_group_static {
+public:
+  __SYCL_ALWAYS_INLINE work_group_static() = default;
+  work_group_static(const work_group_static &) = delete;
+  work_group_static &operator=(const work_group_static &) = delete;
+
+  operator T &() noexcept { return *getDecorated(); }
+
+  template <class TArg = T>
+  typename std::enable_if<!std::is_array_v<TArg>, work_group_static &>::type
+  operator=(const T &value) noexcept {
+    *getDecorated() = value;
+    return *this;
+  }
+
+  T *operator&() noexcept { return *getDecorated(); }
```

Comment:
1) I don't see how this works. `getDecorated` returns `T *`, then we dereference it, yes the result is still `T *`.
2) I don't like this "smart" overload, but that's a question to the language team. Also tagging @gmlueck 
3) Do we need a `const` overload as well/instead of this one?

Reply:
> 2\. I don't like this "smart" overload, but that's a question to the language team. Also tagging @gmlueck

@Pennycook what do you think?  It seems like we need some what to get a pointer to the underlying `T` object.  If we don't have `operator&`, I guess we'd need to add a member function like `get_pointer`.

Reply:
`operator&()` was part of the original specification design, and we added it in order to maintain the illusion that a `work_group_static` variable is just a variable:

```c++
syclex::work_group_static<int[10]> array;
void kernel() {
  int* ptr = &array; // we need this to work
}
```

It's not clear to me if we still need this definition now that `work_group_static` is no longer a view.  If the compiler is going to make all of this work by putting the `work_group_static` *itself* into work-group local memory, then taking the address of `array` will (probably) return the same value as taking the address of the member variable inside `array` (assuming `work_group_static` is standard layout).

I guess one problem is that without an `operator&` overload, the result is going to be a `work_group_static<T>*` instead of a `T*`.  So maybe what we really want is to define:
```c++
 T* operator&() noexcept { return reinterpret_cast<T*>(this); }
```